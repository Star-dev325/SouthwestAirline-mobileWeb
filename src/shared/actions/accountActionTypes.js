import { actionCreator } from 'src/shared/redux/actionCreator';

const { createApiActions, createTypes } = actionCreator('account');

const types = {
  sync: [
    'SAVE_ACCOUNT_INFO',
    'SAVE_ACCOUNT_NUMBER',
    'SAVE_CORPORATE_INFO',
    'SAVE_RECENT_FLIGHT_DESTINATION',    
    'SET_IS_LOGGED_IN',
    'SET_IS_LOGOUT_PENDING',
    'UPDATE_ACCOUNT_INFO_FETCHED'
  ],
  async: [
    'FETCH_ACCOUNT_INFO',
    'FETCH_SALESFORCE_GUID',
    'FETCH_USER_INFO',
    'GET_SALESFORCE_GUID',
    'GRANT_CORPORATE_TOKEN',
    'GRANT_LEISURE_TOKEN',
    'LOGIN_USER',
    'LOGOUT_USER'
  ]
};

export const apiActionCreator = createApiActions;
export default createTypes(types);