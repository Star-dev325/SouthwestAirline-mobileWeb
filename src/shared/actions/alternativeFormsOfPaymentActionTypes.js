import { actionCreator } from 'src/shared/redux/actionCreator';

const { createTypes, createApiActions } = actionCreator('alternativeFormsOfPayment');

const types = {
  sync: ['SAVE_FORM_DATA', 'INTEGRATION_FAILED', 'RESET_AVAILABILITY'],
  async: [
    'FETCH_AVAILABILITY',
    'UPDATE_AVAILABILITY',
    'INITIATE_PAYMENT',
    'RETRIEVE_PARAMS',
    'RELOAD_AND_SUBMIT',
    'CONFIRM',
    'SEND_ERROR'
  ]
};

export const apiActionCreator = createApiActions;
export default createTypes(types);
