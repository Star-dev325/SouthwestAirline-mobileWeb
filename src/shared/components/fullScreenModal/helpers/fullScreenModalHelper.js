import { history } from 'src/appHistory';
import BrowserObject from 'src/shared/helpers/browserObject';
import { mergeQuery, combineUri, getQueryStringParameterByKey } from 'src/shared/helpers/urlHelper';
import { addEventListenerOnce } from 'src/shared/helpers/eventHelpers';
import Q from 'q';

const { location } = BrowserObject;

export const showFullScreenModal = (id) => {
  if (getQueryStringParameterByKey('_modal') !== id) {
    return Q(history.push(combineUri(location.pathname, mergeQuery({ _modal: id }, location.search))));
  } else {
    return Q();
  }
};

export const hideFullScreenModal = (id) => {
  if (getQueryStringParameterByKey('_modal') === id) {
    const deferred = Q.defer();

    addEventListenerOnce(window, 'popstate', () => {
      deferred.resolve();
    });
    history.goBack();

    return deferred.promise;
  } else {
    return Q();
  }
};

export const hideModalAndUpdateFormField = (id, fieldName, fieldValue, onChange) => {
  hideFullScreenModal(id);
  onChange(fieldName, fieldValue);
};

export const getModalId = () => getQueryStringParameterByKey('_modal');
