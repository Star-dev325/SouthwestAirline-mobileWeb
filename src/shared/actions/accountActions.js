// @flow
import { removeCookies as removeResponsiveCookies, removeResponsiveStores } from '@swa-ui/authentication/logoutHandler';
import i18n from '@swa-ui/locale';
import _ from 'lodash';
import store from 'store2';
import { cleanUpFrequentTravelerSelected } from 'src/airBooking/actions/airBookingActions';
import * as ChaseActions from 'src/chase/actions/chaseActions';
import CheckInLocalStorageHelper from 'src/checkIn/helpers/checkInLocalStorageHelper';
import transformToAccountInfo from 'src/myAccount/transformers/accountInfoTransformer';
import AccountActionTypes, { apiActionCreator } from 'src/shared/actions/accountActionTypes';
import * as CreditCardActions from 'src/shared/actions/creditCardActions';
import * as FormDataActions from 'src/shared/actions/formDataActions';
import * as AccountsApi from 'src/shared/api/accountsApi';
import * as OAuthApi from 'src/shared/api/oAuthApi';
import * as SalesforceApi from 'src/shared/api/salesforceApi';
import { deleteUserInfo, loadUserInfo, saveUserInfo } from 'src/shared/cache/localStorageCache';
import { ERROR_CID_NOT_AVAILABLE } from 'src/shared/constants/errorCodes';
import * as AccountInfoHelper from 'src/shared/helpers/accountInfoHelper';
import { containsApiErrorCodes } from 'src/shared/helpers/errorCodesHelper';
import { decodeJwt, hasActiveSessionCookies } from 'src/shared/helpers/loginSessionHelper';
import StorageKeys from 'src/shared/helpers/storageKeys';
import * as Cookie from 'src/shared/swa-persistence/cookie';
import { transformToCorporateInfo } from 'src/shared/transformers/corporateInfoTransformer';
import { transformToOAuthErrorMessage } from 'src/shared/transformers/oauthErrorMessageTransformer';
import { transformUserInfo } from 'src/shared/transformers/userInfoTransformer';
import { expireHomeNavMenu } from 'src/wcm/actions/wcmActions';

import type { Company } from 'src/airBooking/flow-typed/airBooking.types';
import type { LoginFormDataType } from 'src/login/flow-typed/login.types';
import { LOGIN_RR_MISMATCH_ERROR_CODE } from 'src/shared/helpers/validateLoginHelper';

const {
  ACCOUNT__FETCH_ACCOUNT_INFO,
  ACCOUNT__FETCH_SALESFORCE_GUID,
  ACCOUNT__FETCH_USER_INFO,
  ACCOUNT__GET_SALESFORCE_GUID,
  ACCOUNT__GRANT_CORPORATE_TOKEN,
  ACCOUNT__GRANT_LEISURE_TOKEN,
  ACCOUNT__LOGIN_USER,
  ACCOUNT__LOGOUT_USER,
  ACCOUNT__SAVE_ACCOUNT_INFO,
  ACCOUNT__SAVE_ACCOUNT_NUMBER,
  ACCOUNT__SAVE_CORPORATE_INFO,
  ACCOUNT__SAVE_RECENT_FLIGHT_DESTINATION,
  ACCOUNT__SET_IS_LOGGED_IN,
  ACCOUNT__UPDATE_ACCOUNT_INFO_FETCHED
} = AccountActionTypes;

const { loginUser, loginUserSuccess, loginUserFailed } = apiActionCreator(ACCOUNT__LOGIN_USER);

export const login = (loginRequest: LoginFormDataType) => (dispatch: *, getState: () => *) => {
  dispatch(loginUser());
  const state = _.cloneDeep(getState());
  const CORPORATE_INFO_TIMEOUT_MIN = _.get(
    state,
    'app.wcmContent.applicationProperties.CORPORATE_INFO_TIMEOUT_MIN',
    '30'
  );
  const VALIDATE_LOGIN_USER = _.get(state, 'app.toggles.VALIDATE_LOGIN_USER', false);

  store.set(StorageKeys.VALIDATE_LOGIN_USER, VALIDATE_LOGIN_USER);

  return OAuthApi.login(loginRequest)
    .then((oauthLoginResponse) => {
      const response = _getDecodedOauthResponse(oauthLoginResponse);

      AccountInfoHelper.rememberOauthLoginSession(response);
      const { shouldRememberUser, userNameOrAccountNumber } = loginRequest;

      if (shouldRememberUser) {
        AccountInfoHelper.rememberUser(userNameOrAccountNumber);
      } else {
        AccountInfoHelper.forgetUser();
      }
      CheckInLocalStorageHelper.clearAcceptedHazmatDeclarations();
      const corporateInfo = transformToCorporateInfo(response, CORPORATE_INFO_TIMEOUT_MIN);

      if (corporateInfo) {
        dispatch(updateCorporateInfo(corporateInfo));
      }
    })
    .then(() => dispatch(getAccountInfo()))
    .then((transformedAccountInfo) => dispatch(loginUserSuccess(transformedAccountInfo)))
    .then(() => dispatch(expireHomeNavMenu()))
    .catch((err) => {
      dispatch(loginUserFailed(transformToOAuthErrorMessage(err)));

      LOGIN_RR_MISMATCH_ERROR_CODE === err?.responseJSON?.code && dispatch(logout());

      throw err;
    });
};

const { fetchAccountInfo, fetchAccountInfoSuccess, fetchAccountInfoFailed } =
  apiActionCreator(ACCOUNT__FETCH_ACCOUNT_INFO);

export const getAccountInfo = () => (dispatch: *, getState: () => *) => {
  const state = _.cloneDeep(getState());
  const ACCOUNT_INFO_TIMEOUT_MIN = _.get(state, 'app.wcmContent.applicationProperties.ACCOUNT_INFO_TIMEOUT_MIN', '1');

  dispatch(fetchAccountInfo());

  return AccountsApi.fetchAccountInfo()
    .then((response) => {
      dispatch(fetchAccountInfoSuccess(response));

      const transformedAccountInfo = transformToAccountInfo(response, ACCOUNT_INFO_TIMEOUT_MIN);

      AccountInfoHelper.rememberBasicAccountInfo(transformedAccountInfo);

      dispatch(updateAccountInfo(transformedAccountInfo));
      dispatch(setIsLoggedIn(true));
      dispatch(saveAccountNumber(_.get(transformedAccountInfo, 'customerInfo.accountNumber')));
      dispatch(updateAccountInfoFetched(true));

      return transformedAccountInfo;
    })
    .catch((err) => dispatch(fetchAccountInfoFailed(err)));
};

export const refreshCorporateInfo =
  (isHardRefresh: boolean = false) =>
    (dispatch: *, getState: () => *) => {
      const state = getState();
      const localStorageLoginInfo = store.get(StorageKeys.OAUTH_LOGIN_STATUS) || {};
      const activeCompanyIdAssociations = state?.app?.account?.corporateInfo?.activeCompanyIdAssociations;
      const shouldRefreshCorporateInfo =
      !activeCompanyIdAssociations &&
      (
        (localStorageLoginInfo &&
          localStorageLoginInfo['corporate.customerUserInformation.activeCompanyIdAssociations']) ||
        []
      ).length > 0;

      if (shouldRefreshCorporateInfo) {
        isHardRefresh
          ? deleteUserInfo()
          : _.includes(localStorageLoginInfo.scope, 'swabiz') && dispatch(removeSelectedCompany());
        !_.includes(localStorageLoginInfo.scope, 'swabiz') && dispatch(getUserInfo());
      }
    };

const { fetchUserInfo, fetchUserInfoSuccess, fetchUserInfoFailed } = apiActionCreator(ACCOUNT__FETCH_USER_INFO, {
  isSpinnerNeeded: false,
  shouldHideError: true
});

export const getUserInfo =
  (shouldUseFreshData: boolean = false) =>
    (dispatch: *, getState: () => *) => {
      const state = getState();
      const cachedUserInfo = loadUserInfo();

      if (!_.get(state, 'app.account.isLoggedIn') || !hasActiveSessionCookies()) {
        return Promise.resolve();
      }

      if (cachedUserInfo && !shouldUseFreshData) {
        return cachedUserInfo;
      }

      dispatch(fetchUserInfo());

      return OAuthApi.fetchUserInfo()
        .then((response) => {
          const transformedUserInfo = transformUserInfo(response);

          dispatch(fetchUserInfoSuccess(transformedUserInfo));
          const CORPORATE_INFO_TIMEOUT_MIN = _.get(
            state,
            'app.wcmContent.applicationProperties.CORPORATE_INFO_TIMEOUT_MIN',
            '30'
          );
          const isWebView = _.get(state, 'app.webView.isWebView', false);
          const selectedCompany = _.get(state, 'app.account.corporateInfo.selectedCompany');
          const corporateInfo = transformToCorporateInfo(
            response,
            CORPORATE_INFO_TIMEOUT_MIN,
            isWebView,
            selectedCompany
          );

          corporateInfo && dispatch(updateCorporateInfo(corporateInfo));
          saveUserInfo(transformedUserInfo);

          return transformedUserInfo;
        })
        .catch((err) => dispatch(fetchUserInfoFailed(err)));
    };

const { fetchSalesforceGuid, fetchSalesforceGuidSuccess, fetchSalesforceGuidFailed } = apiActionCreator(ACCOUNT__FETCH_SALESFORCE_GUID);

export const getSalesforceGuid = () => (dispatch: *) => {
  dispatch(fetchSalesforceGuid());

  return SalesforceApi.getSalesforceGuid().then((response) => {
    const { guidNumber } = response;
    
    dispatch(setSalesforceGuid(guidNumber));
    dispatch(fetchSalesforceGuidSuccess());
  }).catch(() => {
    dispatch(fetchSalesforceGuidFailed());
  });
};

const { logoutUser, logoutUserSuccess, logoutUserFailed } = apiActionCreator(ACCOUNT__LOGOUT_USER);

export const logout = () => (dispatch: *) => {
  dispatch(logoutUser());

  return OAuthApi.logout()
    .then((res) => {
      dispatch(cleanUpEndOfSession());
      CheckInLocalStorageHelper.clearAcceptedHazmatDeclarations();
      dispatch(logoutUserSuccess(res));
    })
    .catch(() => {
      dispatch(cleanUpEndOfSession());
      CheckInLocalStorageHelper.clearAcceptedHazmatDeclarations();
      dispatch(logoutUserFailed());
    });
};

const { grantCorporateToken, grantCorporateTokenSuccess, grantCorporateTokenFailed } =
  apiActionCreator(ACCOUNT__GRANT_CORPORATE_TOKEN);

export const saveSelectedCompany =
  (selectedCompany: Company) =>
    (dispatch: *, getState: () => *): Promise<*> =>
      new Promise((resolve: *, reject: *) => {
        dispatch(grantCorporateToken());
        const state = _.cloneDeep(getState());
        const CORPORATE_INFO_TIMEOUT_MIN = _.get(
          state,
          'app.wcmContent.applicationProperties.CORPORATE_INFO_TIMEOUT_MIN',
          '30'
        );

        return OAuthApi.grantCorporateToken(selectedCompany.companyId)
          .then((oauthLoginResponse) => {
            const response = _getDecodedOauthResponse(oauthLoginResponse);

            dispatch(grantCorporateTokenSuccess(response));
            AccountInfoHelper.rememberOauthLoginSession(response);
            const corporateInfo = transformToCorporateInfo(response, CORPORATE_INFO_TIMEOUT_MIN);

            corporateInfo && dispatch(updateCorporateInfo({ ...corporateInfo, selectedCompany }));
            resolve();
          })
          .catch((err) => {
            const companyDeactivatedMessage = i18n('ERROR__COMPANY_DEACTIVATED');
            const error = err.disableRetry ? { stopNextAction: true } : err;

            if (containsApiErrorCodes(err, ERROR_CID_NOT_AVAILABLE) && companyDeactivatedMessage) {
              _.set(err, 'responseJSON.message', companyDeactivatedMessage);
            }

            err.disableRetry ? dispatch(grantCorporateTokenFailed()) : dispatch(grantCorporateTokenFailed(err));
            reject(error);
          });
      });

const _getDecodedOauthResponse = (oauthLoginResponse: *) => {
  const id_token = Cookie.getValue('id_token');
  const decodedToken = id_token && decodeJwt(id_token);

  return decodedToken ? { ...oauthLoginResponse, ...decodedToken, id_token } : oauthLoginResponse;
};

export const removeSelectedCompany =
  (clearSessionOnFail: boolean = true) =>
    (dispatch: *) =>
      dispatch(handleLeisureTokenExchange(clearSessionOnFail));

const { grantLeisureToken, grantLeisureTokenSuccess, grantLeisureTokenFailed } =
  apiActionCreator(ACCOUNT__GRANT_LEISURE_TOKEN);

export const handleLeisureTokenExchange =
  (clearSessionOnFail: boolean = true) =>
    (dispatch: *, getState: () => *) => {
      const state = _.cloneDeep(getState());
      const CORPORATE_INFO_TIMEOUT_MIN = _.get(
        state,
        'app.wcmContent.applicationProperties.CORPORATE_INFO_TIMEOUT_MIN',
        '30'
      );
      const isWebView = _.get(state, 'app.webView.isWebView');

      if (isWebView) {
        return Promise.resolve();
      } else {
        dispatch(grantLeisureToken());

        return OAuthApi.grantLeisureToken()
          .then((oauthLoginResponse) => {
            const response = _getDecodedOauthResponse(oauthLoginResponse);

            dispatch(grantLeisureTokenSuccess(response));
            AccountInfoHelper.rememberOauthLoginSession(response);
            dispatch(cleanUpCorporateInfo());
            const corporateInfo = transformToCorporateInfo(response, CORPORATE_INFO_TIMEOUT_MIN);

            dispatch(updateCorporateInfo(corporateInfo));
          })
          .catch((err) => {
            if (clearSessionOnFail) {
              CheckInLocalStorageHelper.clearAcceptedHazmatDeclarations();
              dispatch(cleanUpEndOfSession());
            }

            dispatch(grantLeisureTokenFailed(err));
          });
      }
    };

export const cleanUpAccountInfo = () => (dispatch: *) => {
  dispatch(updateAccountInfo(null));
  dispatch(setIsLoggedIn(false));
  AccountInfoHelper.removeAccountInfo();
  dispatch(updateAccountInfoFetched(false));
};

export const cleanUpAssociatedInfo = () => (dispatch: *) => {
  dispatch(CreditCardActions.resetSavedCreditCards());
  dispatch(ChaseActions.resetChaseData());
};

export const cleanUpCorporateInfo = () => (dispatch: *) => {
  Cookie.deleteCookie('corporateToken');
  dispatch(updateCorporateInfo(null));
};

export const cleanUpEndOfSession =
  (isReLogin: boolean = false) =>
    (dispatch: *) => {
      AccountInfoHelper.removeOauthLoginSession();
      dispatch(cleanUpAccountInfo());
      dispatch(cleanUpCorporateInfo());
      dispatch(cleanUpAssociatedInfo());
      dispatch(expireHomeNavMenu());

      if (!isReLogin) {
        dispatch(FormDataActions.resetFormData());
        dispatch(cleanUpFrequentTravelerSelected());
      }

      removeResponsiveStores();
      removeResponsiveCookies();
      deleteUserInfo();
    };

export const saveAccountNumber = (accountNumber: string) => ({
  type: ACCOUNT__SAVE_ACCOUNT_NUMBER,
  accountNumber
});

export const saveRecentFlightAirport = (airportName: string) => ({
  type: ACCOUNT__SAVE_RECENT_FLIGHT_DESTINATION,
  airportName
});

export const setIsLoggedIn = (isLoggedIn: boolean) => ({
  type: ACCOUNT__SET_IS_LOGGED_IN,
  isLoggedIn
});

export const setSalesforceGuid = (salesforceGuid: string) => ({
  type: ACCOUNT__GET_SALESFORCE_GUID,
  salesforceGuid
});

export const updateAccountInfo = (accountInfo: *) => ({
  type: ACCOUNT__SAVE_ACCOUNT_INFO,
  accountInfo
});

export const updateAccountInfoFetched = (isFetched: boolean = false) => ({
  type: ACCOUNT__UPDATE_ACCOUNT_INFO_FETCHED,
  isFetched
});

export const updateCorporateInfo = (corporateInfo: *) => ({
  type: ACCOUNT__SAVE_CORPORATE_INFO,
  corporateInfo
});