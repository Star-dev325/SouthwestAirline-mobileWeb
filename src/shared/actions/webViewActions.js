// @flow
import dayjs from 'dayjs';
import _ from 'lodash';
import { resetAirBookingPurchaseData } from 'src/airBooking/actions/airBookingActions';
import * as CarBookingActions from 'src/carBooking/actions/carBookingActions';
import {
  clearSpinnerAndModalOptions,
  setReLoginCallbackFunctions,
  setRetryFunctions
} from 'src/login/actions/reLoginModalActions';
import {
  cleanUpAccountInfo,
  cleanUpAssociatedInfo,
  getAccountInfo,
  updateAccountInfo,
  updateCorporateInfo
} from 'src/shared/actions/accountActions';
import { initiateVoidTransaction } from 'src/shared/actions/alternativeFormsOfPaymentActions';
import { asyncActionFinish, asyncActionStart } from 'src/shared/actions/sharedActions';
import WebViewActionTypes, { apiActionCreator } from 'src/shared/actions/webViewActionTypes';
import * as LocalStorageCache from 'src/shared/cache/localStorageCache';
import { PAYMENT_METHODS } from 'src/shared/constants/alternativeFormsOfPaymentConstants';
import { APPLE_PAY_CARD_ID } from 'src/shared/constants/creditCardConstants';
import { LOGIN_STATES } from 'src/shared/constants/webViewConstants';
import * as AccountInfoHelper from 'src/shared/helpers/accountInfoHelper';
import BrowserObject from 'src/shared/helpers/browserObject';
import { decodeJwt } from 'src/shared/helpers/loginSessionHelper';
import { forceDomUpdatesAndThenExecute } from 'src/shared/helpers/uiHelper';
import * as WebViewHelper from 'src/shared/helpers/webViewHelper';
import * as Cookie from 'src/shared/swa-persistence/cookie';
import { transformToCorporateInfo } from 'src/shared/transformers/corporateInfoTransformer';

const {
  WEB_VIEW__HANDLE_APPLE_PAY,
  WEB_VIEW__HANDLE_AUTH_EVENT,
  WEB_VIEW__HANDLE_DEEP_LINK_CONTINUE,
  WEB_VIEW__HANDLE_EXTERNAL_PAYMENT_AUTHORIZED,
  WEB_VIEW__HANDLE_OAUTH,
  WEB_VIEW__HANDLE_PAYPAL_AUTH,
  WEB_VIEW__HANDLE_ROUTE_CHANGE,
  WEB_VIEW__SEND_CHASE_SESSION,
  WEB_VIEW__SEND_EXIT,
  WEB_VIEW__SEND_ENABLE_NAVIGATION_CONTROLS,
  WEB_VIEW__SEND_DISPLAY_APP_REVIEW,
  WEB_VIEW__SEND_DISPLAY_LOGIN,
  WEB_VIEW__SEND_SHARE_FLIGHT_STATUS_DETAILS,
  WEB_VIEW__SET_ADOBE_ID,
  WEB_VIEW__SET_DEVICE_TYPE,
  WEB_VIEW__SET_IS_NOT_WEB_VIEW,
  WEB_VIEW__SET_IS_WEB_VIEW,
  WEB_VIEW__SET_WEB_VIEW_CHANNEL,
  WEB_VIEW__SET_SHARE_FLIGHT_STATUS,
  WEB_VIEW__SET_UPGRADE_TYPE,
  WEB_VIEW__UPDATE_ACCOUNT
} = WebViewActionTypes;
const { location } = BrowserObject;

import type { RouterHistory } from 'react-router';
import type {
  NativeAppLoginOptions,
  Dispatch as ThunkDispatch,
  WebViewOAuth
} from 'src/shared/flow-typed/shared.types';

const { NATIVE_LOG_IN, NATIVE_LOG_OUT } = LOGIN_STATES;

export const isWebView = () => ({
  type: WEB_VIEW__SET_IS_WEB_VIEW
});

export const isNotWebView = () => ({
  type: WEB_VIEW__SET_IS_NOT_WEB_VIEW
});

export const exitWebView = (route: string = '') => ({
  type: WEB_VIEW__SEND_EXIT,
  route
});

export const setDeviceType = (value: string) => ({
  type: WEB_VIEW__SET_DEVICE_TYPE,
  value
});

export const setWebViewChannel = (value: string) => ({
  type: WEB_VIEW__SET_WEB_VIEW_CHANNEL,
  value
});

export const setAdobeId = (value: string) => ({
  type: WEB_VIEW__SET_ADOBE_ID,
  value
});

export const setShareFlightStatus = (value: boolean) => ({
  type: WEB_VIEW__SET_SHARE_FLIGHT_STATUS,
  value
});

export const shareFlightStatusDetails = (shareDetails: *) => ({
  type: WEB_VIEW__SEND_SHARE_FLIGHT_STATUS_DETAILS,
  shareDetails
});

export const setUpgradeType = (value: string) => ({
  type: WEB_VIEW__SET_UPGRADE_TYPE,
  value
});

export const handleOAuth = (isAdd: boolean, value: WebViewOAuth) => ({
  type: WEB_VIEW__HANDLE_OAUTH,
  value,
  isAdd
});

export const handleAuthEvent = (value: string) => ({
  type: WEB_VIEW__HANDLE_AUTH_EVENT,
  value
});

export const handleRouteChange = (history: RouterHistory, route: string, state: string) => ({
  type: WEB_VIEW__HANDLE_ROUTE_CHANGE,
  history,
  route,
  state
});

export const showNativeAppLogin = (options: NativeAppLoginOptions = {}) => ({
  type: WEB_VIEW__SEND_DISPLAY_LOGIN,
  options
});

export const handleNativeLogout = () => (dispatch: ThunkDispatch, getState: *) => {
  const webView = _.get(getState(), 'app.webView.isWebView');

  webView && WebViewHelper.handleNativeLogout();
};

export const enableNavigationControls = (isEnabled: boolean) => ({
  type: WEB_VIEW__SEND_ENABLE_NAVIGATION_CONTROLS,
  isEnabled
});

export const displayAppReview = () => ({
  type: WEB_VIEW__SEND_DISPLAY_APP_REVIEW
});

export const setPaypalAuthorized = (token: string) => ({
  type: WEB_VIEW__HANDLE_PAYPAL_AUTH,
  token
});

export const handleNativeApplePay = (request: string) => ({
  type: WEB_VIEW__HANDLE_APPLE_PAY,
  request
});

export const handleChaseSession = (value: string) => ({
  type: WEB_VIEW__SEND_CHASE_SESSION,
  value
});

export const handleDeepLinkContinue = (value: boolean) => ({
  type: WEB_VIEW__HANDLE_DEEP_LINK_CONTINUE,
  value
});

export const handleSaveChaseOffers = (encodedResponse: *) => () => {
  const response = WebViewHelper.decodeMessage(encodedResponse);
  const expirationTimestamp = _.get(response, 'expirationTimestamp');

  LocalStorageCache.saveChasePrequalOffers(response, null, expirationTimestamp);
};

export const handleExternalPaymentAuthorized = (value: ?string) => ({
  type: WEB_VIEW__HANDLE_EXTERNAL_PAYMENT_AUTHORIZED,
  value
});

const saveAccountInfo = () => (dispatch: ThunkDispatch) => {
  if (!AccountInfoHelper.getOauthLoginSession()) {
    return Promise.resolve();
  }

  return dispatch(getAccountInfo()).then((accountInfo) => {
    dispatch(CarBookingActions.saveUserAccountInfo(accountInfo));
  });
};

export const refreshAuth = () => (dispatch: ThunkDispatch) => {
  const existingAuth = AccountInfoHelper.getOauthLoginSession() || {};
  const accountInfo = AccountInfoHelper.getAccountInfo();
  const shouldSaveAccountInfo =
    dayjs().isAfter(_.get(accountInfo, 'expirationDate')) || (!_.isEmpty(existingAuth) && _.isEmpty(accountInfo));

  shouldSaveAccountInfo && dispatch(saveAccountInfo());
  !shouldSaveAccountInfo && !_.isEmpty(accountInfo) && dispatch(updateAccountInfo(accountInfo));
};

const { updateAccount, updateAccountSuccess, updateAccountFailed } = apiActionCreator(WEB_VIEW__UPDATE_ACCOUNT);

export const addOAuth = (encodedCredentials: string, getState: () => *) => (dispatch: ThunkDispatch) => {
  const id_token = Cookie.getValue('id_token');
  const decodedIdToken = id_token ? decodeJwt(id_token) : {};
  const decodedAuth = { ...decodedIdToken, id_token };

  if (!_.isEmpty(decodedIdToken)) {
    dispatch(updateAccount());
    AccountInfoHelper.rememberOauthLoginSession(decodedAuth);

    const CORPORATE_INFO_TIMEOUT_MIN = _.get(
      _.cloneDeep(getState()),
      'app.wcmContent.applicationProperties.CORPORATE_INFO_TIMEOUT_MIN',
      '30'
    );
    const corporateInfo = transformToCorporateInfo(decodedAuth, CORPORATE_INFO_TIMEOUT_MIN, true);

    dispatch(updateCorporateInfo(corporateInfo));

    forceDomUpdatesAndThenExecute(() => {
      dispatch(saveAccountInfo())
        .then(() => dispatch(updateAccountSuccess(NATIVE_LOG_IN)))
        .then(() => !encodedCredentials && dispatch(handleReLogin()))
        .catch(() => dispatch(updateAccountFailed()));
    });
  }
};

export const handleReLogin = () => (dispatch: *, getState: *) => {
  const state = _.cloneDeep(getState());
  const { reLoginModal = {} } = state?.app;
  const { reLoginCallbackFunctions = {}, reLoginLocation, retryFunctions = [] } = reLoginModal;
  const { postLoginCallbackFn = _.noop } = reLoginCallbackFunctions || {};
  const hasCallbackFunctionsForCurrentPage = reLoginLocation === location.pathname;

  !hasCallbackFunctionsForCurrentPage && dispatch(setReLoginCallbackFunctions({}));
  retryFunctions.length > 0 &&
    dispatch(_retryFailedCalls())
      .then(postLoginCallbackFn)
      .finally(() => dispatch(setReLoginCallbackFunctions({})));
};

const _retryFailedCalls = () => (dispatch: *, getState: *) => {
  const state = _.cloneDeep(getState());
  const { reLoginModal = {} } = state?.app;
  const { retryFunctions = [] } = reLoginModal;

  dispatch(asyncActionStart());
  Promise.all(_.map(retryFunctions, (retryFunction) => retryFunction()))
    .catch(_.noop)
    .finally(() => {
      dispatch(setRetryFunctions([]));
      dispatch(asyncActionFinish());
    });
};

export const removeOAuth = () => (dispatch: ThunkDispatch, getState: () => *) => {
  const existingAuth = AccountInfoHelper.getOauthLoginSession() || {};
  const hasToken = !!existingAuth.id_token;
  const state = _.cloneDeep(getState());
  const { reLoginModal = {} } = _.get(state, 'app');
  const { reLoginCallbackFunctions = {}, reLoginLocation } = reLoginModal;
  const { continueAsGuestFn } = reLoginCallbackFunctions || {};
  const hasGuestFlowForCurrentPage = reLoginLocation === location.pathname;

  if (hasToken) {
    dispatch(updateAccount());
    AccountInfoHelper.removeOauthLoginSession();
    forceDomUpdatesAndThenExecute(() => {
      dispatch(resetAirBookingPurchaseData());
      dispatch(cleanUpAccountInfo());
      dispatch(cleanUpAssociatedInfo());
      hasGuestFlowForCurrentPage && continueAsGuestFn && continueAsGuestFn(true);
      dispatch(clearSpinnerAndModalOptions());
      dispatch(updateAccountSuccess(NATIVE_LOG_OUT));
    });
  }
};

export const voidApplePayTransaction = () => (dispatch: ThunkDispatch, getState: () => *) => {
  const state = _.cloneDeep(getState());
  const applePayCard = state?.app?.applePay?.applePayCard;
  const paymentInfo = applePayCard?.formData?.paymentInfo;
  const hasSelectedApplePay = !!paymentInfo && paymentInfo.selectedCardId === APPLE_PAY_CARD_ID;

  if (hasSelectedApplePay) {
    const CEPTOR_VOID_API = state?.app?.toggles?.CEPTOR_VOID_API;
    const { reLoginModal = {} } = state?.app || {};
    const { reLoginLocation = "" } = reLoginModal || {};
    const hasGuestFlowForCurrentPage = reLoginLocation === location.pathname;
    const isValidApplePayCard = applePayCard && Object.keys(applePayCard).length > 0;

    if (CEPTOR_VOID_API && hasGuestFlowForCurrentPage && isValidApplePayCard) {
      dispatch(initiateVoidTransaction(PAYMENT_METHODS.APPLE_PAY, null, true, 'user continued as guest'));
    }
  }
};
