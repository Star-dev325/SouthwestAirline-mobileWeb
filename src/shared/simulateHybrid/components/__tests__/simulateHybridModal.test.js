jest.mock('src/shared/components/fullScreenModal/helpers/fullScreenModalHelper', () => ({
  hideFullScreenModal: jest.fn()
}));
jest.mock('src/shared/actions/webViewActions', () => ({
  isNotWebView: jest.fn().mockReturnValue({ type: 'isNotWebView' }),
  isWebView: jest.fn().mockReturnValue({ type: 'isWebView' })
}));

import { fireEvent } from '@testing-library/react';
import { isNotWebView, isWebView } from 'src/shared/actions/webViewActions';
import { hideFullScreenModal } from 'src/shared/components/fullScreenModal/helpers/fullScreenModalHelper';
import SimulateHybridModal from 'src/shared/simulateHybrid/components/simulateHybridModal';
import { integrationRender } from 'test/unit/helpers/testingLibraryUtils';

describe('SimulateHybridModal', () => {
  it('should render correctly', () => {
    const { container } = createComponent();

    expect(container).toMatchSnapshot();
  });

  it('should dismiss the full screen modal when the done button is clicked', () => {
    const { getByText } = createComponent();

    fireEvent.click(getByText('Done'));

    expect(hideFullScreenModal).toHaveBeenCalled();
  });

  it('should update web view state when the checkbox is turned on', () => {
    const { getByText } = createComponent(); 

    fireEvent.click(getByText('Enable Hybrid'));

    expect(isWebView).toHaveBeenCalled();
  });

  it('should update web view state when the checkbox is turned on', () => {
    const { getByText } = createComponent({ app: { webView: { isWebView: true } } });

    fireEvent.click(getByText('Enable Hybrid'));

    expect(isNotWebView).toHaveBeenCalled();
  });

  const defaultState = {
    app: {
      webView: {
        isWebView: false
      }
    },
    router: {
      location: {
        search: {
          activeIdInURL: '_modal=_hybrid'
        }
      }
    }
  };
  const createComponent = (state = {}) => {
    const mergedState = {
      ...defaultState,
      ...state
    };

    return integrationRender()(mergedState, SimulateHybridModal, {});
  };
});