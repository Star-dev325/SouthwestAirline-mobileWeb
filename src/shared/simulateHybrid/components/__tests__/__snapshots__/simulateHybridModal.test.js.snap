// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SimulateHybridModal should render correctly 1`] = `
<div>
  <div>
    <div />
  </div>
</div>
`;
